File: EXPERIENCIA-DE-PRÁCTICA-4-USO-DE-OBJETOS-colaboración/EJERCICIO-4.2/Clase_Producto.java
import java.util.ArrayList; 

import java.util.List; 

 

// Clase Producto: Representa un producto con atributos como nombre, precio y cantidad. 

class Producto { 

    // Atributos 

    private String nombre; 

    private double precio; 

    private int cantidad; 

 

    // Constructor 

    public Producto(String nombre, double precio, int cantidad) { 

        this.nombre = nombre; 

        this.precio = precio; 

        this.cantidad = cantidad; 

    } 

 

    // Métodos set para precio y cantidad 

    public void setPrecio(double precio) { 

        this.precio = precio; 

    } 

 

    public void setCantidad(int cantidad) { 

        this.cantidad = cantidad; 

    } 

 

    // Getters para obtener el nombre, precio y cantidad del producto 

    public String getNombre() { 

        return nombre; 

    } 

 

    public double getPrecio() { 

        return precio; 

    } 

 

    public int getCantidad() { 

        return cantidad; 

    } 

} 

 

// Clase Factura: Representa una factura con un método para agregar productos. 

class Factura { 

    // Lista de productos 

    private List<Producto> productos; 

 

    // Constructor 

    public Factura() { 

        this.productos = new ArrayList<>(); 

    } 

 

    // Método para agregar un producto a la factura 

    public void agregarProducto(Producto producto) { 

        productos.add(producto); 

    } 

 

    // Método para calcular el total de la factura 

    public double calcularTotal() { 

        double total = 0; 

        for (Producto producto : productos) { 

            total += producto.getPrecio() * producto.getCantidad(); 

        } 

        return total; 

    } 

} 

 

public class Main { 

    public static void main(String[] args) { 

        // Crear algunos productos 

        Producto producto1 = new Producto("Producto 1", 10.0, 2); 

        Producto producto2 = new Producto("Producto 2", 20.0, 1); 

 

        // Crear una factura y agregar productos 

        Factura factura = new Factura(); 

        factura.agregarProducto(producto1); 

        factura.agregarProducto(producto2); 

 

        // Calcular el total de la factura 

        double totalFactura = factura.calcularTotal(); 

        System.out.println("Total de la factura: $" + totalFactura); 

    } 

} 

File: EXPERIENCIA-DE-PRÁCTICA-4-USO-DE-OBJETOS-colaboración/EJERCICIO-4.2/ClaseProducto.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Producto {
private:
    string nombre;
    double precio;
    int cantidad;

public:

    // nombre del producto
    void setNombre(string _nombre) {
        nombre = _nombre;
    }

    // precio del producto
    void setPrecio(double _precio) {
        precio = _precio;
    }

    // cantidad del producto
    void setCantidad(int _cantidad) {
        cantidad = _cantidad;
    }


    // obtener el nombre del producto
    string getNombre() const {
        return nombre;
    }

    // obtener el precio del producto
    double getPrecio() const {
        return precio;
    }

    // la cantidad del producto
    int getCantidad() const {
        return cantidad;
    }
};

class Factura {
private:
    vector<Producto> productos;

public:
    // agregar un producto a la factura
    void agregarProducto(Producto &producto) {
        productos.push_back(producto);
    }

    // mostrar los detalles de la factura
    void mostrarDetalles() {
        cout << "Detalles de la Factura:" << endl;
        double total = 0.0;
        for (const auto &producto : productos) {
        cout << "\tProducto" << producto.getNombre() << "\nPrecio: S/" << producto.getPrecio() << "\nCantidad: " << producto.getCantidad() << endl;
            total += producto.getPrecio() * producto.getCantidad();
        }
        cout << "Total: S/" << total << endl;
    }
};

int main() {
    Producto producto1, producto2;

    // atributos para el producto 1
    producto1.setNombre("Azucar");
    producto1.setPrecio(10.60);
    producto1.setCantidad(2);

    // atributos para el producto 2
    producto2.setNombre("Pan");
    producto2.setPrecio(2);
    producto2.setCantidad(8);

    Factura factura;

    // Agregar productos a la factura
    factura.agregarProducto(producto1);
    factura.agregarProducto(producto2);

    // Mostrar los detalles de la factura
    factura.mostrarDetalles();

    return 0;
}
